src/lib/rnp.c

```diff
@@ -1084,85 +1084,71 @@ rnp_public_count(rnp_t *rnp)
-    pgp_key_t *          subkey_sec = NULL;
-    pgp_key_t *          subkey_pub = NULL;
+    pgp_key_t            primary_sec = {0};
+    pgp_key_t            primary_pub = {0};
+    pgp_key_t            subkey_sec = {0};
+    pgp_key_t            subkey_pub = {0};
     char *               cp = NULL;
-    bool                 ok = false;
     key_store_format_t   key_format = ((rnp_key_store_t *) rnp->secring)->format;
 
-    primary_sec = calloc(1, sizeof(*primary_sec));
-    primary_pub = calloc(1, sizeof(*primary_pub));
-    subkey_sec = calloc(1, sizeof(*subkey_sec));
-    subkey_pub = calloc(1, sizeof(*subkey_pub));
-    if (!primary_sec || !primary_pub || !subkey_sec || !subkey_pub) {
-        goto end;
-    }
     if (!pgp_generate_keypair(&action->primary.keygen,
                               &action->subkey.keygen,
                               true,
-                              primary_sec,
-                              primary_pub,
-                              subkey_sec,
-                              subkey_pub,
+                              &primary_sec,
+                              &primary_pub,
+                              &subkey_sec,
+                              &subkey_pub,
                               key_format)) {
         RNP_LOG("failed to generate keys");
-        goto end;
+        return false;
     }
 
     // show the primary key
-    pgp_sprint_key(rnp->io, NULL, primary_pub, &cp, "pub", &primary_pub->key.pubkey, 0);
+    pgp_sprint_key(rnp->io, NULL, &primary_pub, &cp, "pub", &primary_pub.key.pubkey, 0);
     (void) fprintf(stdout, "%s", cp);
     free(cp);
 
     // protect the primary key
     if (!pgp_key_protect(
-          primary_sec, key_format, &action->primary.protection, &rnp->passphrase_provider)) {
-        goto end;
+          &primary_sec, key_format, &action->primary.protection, &rnp->passphrase_provider)) {
+        return false;
     }
 
     // show the subkey
-    pgp_sprint_key(rnp->io, NULL, subkey_pub, &cp, "sub", &subkey_pub->key.pubkey, 0);
+    pgp_sprint_key(rnp->io, NULL, &subkey_pub, &cp, "sub", &subkey_pub.key.pubkey, 0);
     (void) fprintf(stdout, "%s", cp);
     free(cp);
 
     // protect the subkey
     if (!pgp_key_protect(
-          subkey_sec, key_format, &action->subkey.protection, &rnp->passphrase_provider)) {
+          &subkey_sec, key_format, &action->subkey.protection, &rnp->passphrase_provider)) {
         RNP_LOG("failed to protect keys");
-        goto end;
+        return false;
     }
 
     // add them all to the key store
-    if (!rnp_key_store_add_key(rnp->io, rnp->secring, primary_sec) ||
-        !rnp_key_store_add_key(rnp->io, rnp->secring, subkey_sec) ||
-        !rnp_key_store_add_key(rnp->io, rnp->pubring, primary_pub) ||
-        !rnp_key_store_add_key(rnp->io, rnp->pubring, subkey_pub)) {
+    if (!rnp_key_store_add_key(rnp->io, rnp->secring, &primary_sec) ||
+        !rnp_key_store_add_key(rnp->io, rnp->secring, &subkey_sec) ||
```

Comment:
This is not about the commit, but about the architecture itself. In my opinion subkey should be part of the key, not the separate entity of the keystore. So, for instance, for decryption/verification operation it is searched by keyid, and for encryption/signing - by user id of the main key. @dewyatt, @flowher, @ronaldtse what do you think?

Reply:
Moved to new issue #502

Reply:
@ronaldtse Thanks. So, I'm merging.